src/sqlfluff/core/linter/runner.py

```diff
@@ -127,54 +133,128 @@ def __init__(self, linter, config, processes) -> None:
         super().__init__(linter, config)
         self.processes = processes
 
-    def run(self, fnames: List[str], fix: bool):
+    def run(self, fnames: List[str], fix: bool) -> Iterator[LintedFile]:
         """Parallel implementation.
 
-        Note that the partials are generated one at a time then
-        passed directly into the pool as they're ready. This means
-        the main thread can do the IO work while passing the parsing
+        Partials are generated one by one. If a file has been successfully
+        rendered, it is set up as an asynchronous job, and if it's been
+        unsuccessful we handle it immediately in the main process. This
+        means the main thread can do the IO work while passing the parsing
         and linting work out to the threads.
+
+        Any completed jobs are always handled before queueing more jobs,
+        and the queue is limited in size, which when reached we simply
+        wait for some progress to be made before enqueueing more work.
+
+        We keep track of the outstanding jobs by keeping a reference to
+        the AsyncResult of each, which is deleted upon receipt of a
+        result from that job.
         """
         with self._create_pool(
             self.processes,
             self._init_global,
             (self.config,),
         ) as pool:
             try:
-                for lint_result in self._map(
```

Comment:
I have an idea for a possible simpler approach (debatably, ha !-- at least less custom code?). Pseudocode:

```
for lint_result in itertools.chain(
    failed_filenames_empty_lint_result(),  # How to implement this? Similar logic as the loop below, I think.)
    self._map(
        pool,
        self._apply,
        self.iter_partials(fnames, fix=fix),  # Need to modify `Iter_partials()` so it skips failed files
    )
):
    # Loop body remains the same as previously
```

Thoughts? I like using `map()` if we can.

Reply:
The problem with `map` is that we don't get _any_ of the results until _all_ of the tasks have completed. The only implementations where we can get the results "as we go" appear to be `imap` and `imap_unordered`, but they have this strange hanging behaviour with exceptions.

I'm all for simple implementations though which take up less space!

Reply:
Oh - and on the skipping failed files - we do still need to return them, they just shouldn't be passed to the workers, so we do kind of need this "two path" handling which I've introduced. I admit that there's some code duplication in there which we can reduce though.